transports/jms/src/main/java/org/mule/transport/jms/JmsMessageDispatcher.java

```diff
@@ -95,186 +98,310 @@ protected boolean isDisableTemporaryDestinations()
-                        return createNullMuleMessage();
+                        }
+                        else
+                        {
+                            return internalBlockingSendAndReceive(producer, consumer, replyTo, jmsMessage, topic, ttl, priority, persistent, timeout);
+                        }
                     }
-                    else
+                    finally
                     {
-                        return createMessageWithJmsMessagePayload(result);
+                        closeConsumer(session, consumer, replyTo);
                     }
                 }
             }
             else
             {
-                // In this case a response was never expected so we return null and not NullPayload.
-                // This generally happens when dispatch is used for an asynchronous endpoint but can also occur when send() is used 
-                // and disableTempDestinations is set.
-                return returnOriginalMessageAsReply ? createMuleMessage(msg) : null;
+                return internalSend(producer, jmsMessage, topic, ttl, priority, persistent);
             }
         }
         finally
         {
             connector.closeQuietly(producer);
-            connector.closeQuietly(consumer);
+            if(!delayedCleanup)
+            {
+                closeSession(session, transacted);
+            }
+        }
+    }
+
+    private String getCorrelationId(MuleEvent event)
+    {
+        if (event.getFlowConstruct() != null)
+        {
+            return event.getFlowConstruct().getMessageInfoMapping().getCorrelationId(event.getMessage());
+        }
+        else
+        {
+            return new MuleMessageInfoMapping().getCorrelationId(event.getMessage());
+        }
+    }
+
+    private MuleMessage internalSend(MessageProducer producer, Message jmsMessage, boolean topic, long ttl, int priority, boolean persistent) throws Exception
+    {
+        connector.getJmsSupport().send(producer, jmsMessage, persistent, priority, ttl, topic, endpoint);
+        return returnOriginalMessageAsReply ? createMuleMessage(jmsMessage) : null;
+    }
+
+    private MuleMessage internalBlockingSendAndAwait(MessageConsumer consumer, MessageProducer producer, Destination replyTo,
+                                                     Message jmsMessage, boolean topic, long ttl, int priority, boolean persistent, int timeout) throws Exception
+    {
+        // need to register a listener for a topic
+        Latch l = new Latch();
+        LatchReplyToListener listener = new LatchReplyToListener(l);
+        consumer.setMessageListener(listener);
+
+        connector.getJmsSupport().send(producer, jmsMessage, persistent, priority, ttl, topic, endpoint);
+
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("Waiting for return event for: " + timeout + " ms on " + replyTo);
+        }
+
+        l.await(timeout, TimeUnit.MILLISECONDS);
```

Comment:
This code is existing code refactored. The message is obtained from the listener after latch.await().  If there is an interuptedException then it will be throw as a runtime exception which will cause a DispatchException to be thrown.